lib/Basic/Diagnostic.cpp

```diff
@@ -0,0 +1,137 @@
+#include "Basic/Diagnostic.hpp"
+
+namespace glu {
+
+void DiagnosticManager::addDiagnostic(
+    DiagnosticSeverity severity, SourceLocation loc, llvm::StringRef message
+)
+{
+    _messages.emplace_back(severity, loc, message.str());
+
+    // Set the error flag if this is an error or fatal error
+    _hasErrors = severity >= DiagnosticSeverity::Error;
```

Comment:
If hasErrors is true and we add a warning, this would set hasErrors back to false